src/core/common/frame_builder.hpp

```diff
@@ -0,0 +1,239 @@
+     * @retval kErrorNoBufs  Insufficient available buffers.
+     *
+     */
+    Error AppendLittleEndianUint32(uint32_t aUint32);
+
+    /**
+     * This method appends bytes from a given buffer to the `FrameBuilder`.
+     *
+     * @param[in] aBuffer    A pointer to a data bytes to append.
+     * @param[in] aLength    Number of bytes in @p aBuffer.
+     *
+     * @retval kErrorNone    Successfully appended the bytes.
+     * @retval kErrorNoBufs  Insufficient available buffers.
+     *
+     */
+    Error AppendBytes(const void *aBuffer, uint16_t aLength);
+
+    /**
+     * This method appends bytes read from a given message to the `FrameBuilder`.
+     *
+     * @param[in] aMessage   The message to read the bytes from.
+     * @param[in] aOffset    The offset in @p aMessage to start reading the bytes from.
+     * @param[in] aLength    Number of bytes to read from @p aMessage and append.
+     *
+     * @retval kErrorNone    Successfully appended the bytes.
+     * @retval kErrorNoBufs  Insufficient available buffers to append the requested @p aLength bytes.
+     * @retval kErrorParse   Not enough bytes in @p aMessage to read @p aLength bytes from @p aOffset.
+     *
+     */
+    Error AppendBytesFromMessage(const Message &aMessage, uint16_t aOffset, uint16_t aLength);
+
+    /**
+     * This method appends an object to the `FrameBuilder`.
+     *
+     * @tparam    ObjectType  The object type to append.
+     *
+     * @param[in] aObject     A reference to the object to append.
+     *
+     * @retval kErrorNone    Successfully appended the object.
+     * @retval kErrorNoBufs  Insufficient available buffers to append @p aObject.
+     *
+     */
+    template <typename ObjectType> Error Append(const ObjectType &aObject)
+    {
+        static_assert(!TypeTraits::IsPointer<ObjectType>::kValue, "ObjectType must not be a pointer");
+
+        return AppendBytes(&aObject, sizeof(ObjectType));
+    }
+
+    /**
+     * This method writes bytes in `FrameBuilder` at a given offset overwriting the previously appended content.
+     *
+     * This method does not perform any bound checks. The caller MUST ensure that the given data length fits within the
+     * previously appended content. Otherwise the behavior of this method is undefined.
+     *
+     * @param[in] aOffset    The offset to begin writing.
+     * @param[in] aBuffer    A pointer to a data buffer to write.
+     * @param[in] aLength    Number of bytes in @p aBuffer.
+     *
+     */
+    void WriteBytes(uint16_t aOffset, const void *aBuffer, uint16_t aLength);
+
+    /**
+     * This methods writes an object to the `FrameBuilder` at a given offset overwriting previously appended content.
+     *
+     * This method does not perform any bound checks. The caller MUST ensure the given data length fits within the
+     * previously appended content. Otherwise the behavior of this method is undefined.
+     *
+     *
+     * @tparam     ObjectType   The object type to write.
```

Comment:
```suggestion
     * previously appended content. Otherwise the behavior of this method is undefined.
     *
     * @tparam     ObjectType   The object type to write.
```